domain/secretbackend/service/service.go

```diff
@@ -0,0 +1,580 @@
+		if backend.Config == nil && len(defaults) > 0 {
+			backend.Config = make(map[string]interface{})
+		}
+		for k, v := range defaults {
+			if _, ok := backend.Config[k]; !ok {
+				backend.Config[k] = v
+			}
+		}
+		err = configValidator.ValidateConfig(nil, backend.Config)
+		if err != nil {
+			return fmt.Errorf("%w: config for provider %q: %w", secretbackenderrors.NotValid, backend.BackendType, err)
+		}
+	}
+	if err := pingBackend(p, backend.Config); err != nil {
+		return errors.Trace(err)
+	}
+
+	var nextRotateTime *time.Time
+	if backend.TokenRotateInterval != nil && *backend.TokenRotateInterval > 0 {
+		if !provider.HasAuthRefresh(p) {
+			return fmt.Errorf("%w: token refresh on secret backend of type %q", secretbackenderrors.NotSupported, p.Type())
+		}
+		nextRotateTime, err = coresecrets.NextBackendRotateTime(s.clock.Now(), *backend.TokenRotateInterval)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	_, err = s.st.CreateSecretBackend(
+		ctx, secretbackend.CreateSecretBackendParams{
+			BackendIdentifier: secretbackend.BackendIdentifier{
+				ID:   backend.ID,
+				Name: backend.Name,
+			},
+			BackendType:         backend.BackendType,
+			TokenRotateInterval: backend.TokenRotateInterval,
+			Config:              convertConfigToString(backend.Config),
+			NextRotateTime:      nextRotateTime,
+		},
+	)
+	if errors.Is(err, domain.ErrDuplicate) {
+		return fmt.Errorf("%w: secret backend with name %q", secretbackenderrors.AlreadyExists, backend.Name)
+	}
+	return errors.Trace(err)
+}
+
+// UpdateSecretBackend updates an existing secret backend.
+func (s *Service) UpdateSecretBackend(ctx context.Context, params UpdateSecretBackendParams) error {
+	if err := params.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+
+	if params.NewName != nil {
+		if err := validateExternalBackendName(*params.NewName); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	// TODO: we should get the latest existing backend, merge the config then validate inside
+	// the update operation transaction.
+	existing, err := s.st.GetSecretBackend(ctx, params.BackendIdentifier)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	p, err := s.registry(existing.BackendType)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	cfgToApply := make(map[string]interface{})
+	for k, v := range existing.Config {
+		cfgToApply[k] = v
+	}
+	for k, v := range params.Config {
+		cfgToApply[k] = v
+	}
+	for _, k := range params.Reset {
+		delete(cfgToApply, k)
+	}
+	configValidator, ok := p.(provider.ProviderConfig)
+	if ok {
+		defaults := configValidator.ConfigDefaults()
+		for _, k := range params.Reset {
+			if defaultVal, ok := defaults[k]; ok {
+				cfgToApply[k] = defaultVal
+			}
+		}
+		err = configValidator.ValidateConfig(convertConfigToAny(existing.Config), cfgToApply)
+		if err != nil {
+			return fmt.Errorf("%w: config for provider %q: %w", secretbackenderrors.NotValid, existing.BackendType, err)
+		}
+	}
+	if !params.Force {
```

Comment:
Ah so force is for whether we ping or not vs validating config. So renaming the params would be good to reflect its purpose.